packages/txservice/src/txservice.ts

```diff
@@ -121,57 +120,84 @@ export class TransactionService {
     const methodId = getUuid();
     this.logger.info({ method, methodId, requestContext, tx }, "Method start");
 
-    const transaction = await Transaction.create(this.logger, this.getProvider(tx.chainId), tx, this.config);
-    let submitFailed = false;
-
+    const transaction = await this.getProvider(tx.chainId).createTransaction(tx);
+    this.logger.debug({ method, methodId, requestContext, params: transaction.params, id: transaction.id }, "Transaction created");
     try {
-      while (!transaction.didFinish()) {
-        // Submit step.
+      while (!transaction.didFinish) {
+        // Submit: send to chain.
         try {
           await this.submitTransaction(transaction, requestContext);
         } catch (error) {
-          // IFF this is the first attempt, throw.
-          // Otherwise, we should go ahead and get the receipt.
-          if (transaction.attempt <= 1) {
+          this.logger.debug(
+            { method, methodId, id: transaction.id },
+            `Transaction received ${error.type} error.`
+          );
+          if (error.type === AlreadyMined.type) {
+            if (transaction.attempt === 1) {
+              // A transaction that's only been attempted once has an expired nonce. This means that TransactionMonitor
+              // assigned us a bunk (already used) nonce.
+
+              // TODO: In this event, we need to go back to the beginning and actually "recreate" the transaction
+              // itself now. Assuming our nonce tracker (TransactionMonitor) is effective, this should normally never occur...
+              // but there is at least 1 legit edge case: if the TransactionMonitor has just come online and is can only rely on
+              // the provider's TransactionCount to assign nonce - and the provider turns out to be incorrect (e.g. off by 1 or 2
+              // pending tx's not in its mempool yet for some reason).
+              // So we may want to replace this throw with a recreate.
+              throw error;
+            } else {
+              // Ignore this error, proceed to validation step.
+              this.logger.debug(
+                { method, methodId, id: transaction.id },
+                "Continuing to confirmation step."
+              );
+            }
+          } else {
             throw error;
-          } else if (error instanceof AlreadyMined) {
-            // If the error is an AlreadyMined error, we probably don't need to log a warning here.
+          }
+        }
+
+        // Validate: wait for 1 confirmation.
+        try {
+          // TODO: Should we emit event to listeners for this step?
```

Comment:
no, we should only emit events when we are certain of the outcome. if we are still waiting for more confirmations before we accept it, we shouldn't emit an external event. if the event is used internally only, thats fine

Reply:
Kk